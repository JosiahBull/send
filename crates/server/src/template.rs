//! Templating for the server, built around handlebars but abstracted away so would be easy to
//! exchange in the future.

use std::num::NonZeroU64;

use handlebars::Handlebars;
use url::Url;

/// [`DownloadPageFields`] contains all of the fields that will be used in the
/// [`../templates/download.hbs`] handlebars template.
#[derive(Debug, serde::Serialize)]
pub struct DownloadPageFields {
    /// The name of the file being downloaded, e.g. `homework.docx`.`
    pub file_name: String,
    /// The full source url where the public key of the user who uploaded this file was acquired.
    /// e.g. `https://www.github.com/josiahbull`, should NOT have a trailing .keys appended.
    pub username_source_url: Url,
    /// The username of the user who uploaded the file, e.g. `josiahbull`.
    pub username: String,
    /// The date the file was uploaded in ms since epoch, e.g. `1620000000000`.
    #[serde(with = "chrono::serde::ts_seconds")]
    pub upload_date: chrono::DateTime<chrono::Utc>,
    /// The date the file will expire in ms since epoch, e.g. `1620000000000`.
    #[serde(with = "chrono::serde::ts_seconds")]
    pub expiry_date: chrono::DateTime<chrono::Utc>,
    /// The size of the file in bytes, e.g. `1024` for a 1kb file.
    pub file_size_bytes: NonZeroU64,
    /// The full URL to download the file, e.g. `https://www.downloads.mysite.com/homework.docx`.
    pub download_url: Url,
    /// The base URL of the site with a trailing slash, e.g. `https://www.downloads.mysite.com`.
    pub base_url: Url,
}

/// A wrapper to handle the templating engine.
#[derive(Debug)]
pub struct Templates {
    /// The handlebars handle.
    handlebars: Handlebars<'static>,
}

impl Templates {
    /// Create a new templating engine with the default configuration.
    pub fn new() -> Self {
        let mut handlebars = Handlebars::new();

        handlebars.set_strict_mode(true);

        handlebars
            .register_template_string("download_page", include_str!("../templates/download.hbs"))
            .expect("Failed to register download_page template");
        Self { handlebars }
    }

    /// Render specifically the download page, with all of the required fields.
    pub fn render_download_page(&self, fields: DownloadPageFields) -> String {
        debug_assert!(
            !fields.username_source_url.as_str().ends_with('/'),
            "username source url should not have trailing slash ="
        );
        debug_assert!(
            !fields.username_source_url.as_str().contains(".keys"),
            "username source url should not end with or contain .keys"
        );
        debug_assert!(
            fields.upload_date < fields.expiry_date,
            "upload date should be before expiry date"
        );
        debug_assert!(
            fields
                .download_url
                .as_str()
                .starts_with(fields.base_url.as_str()),
            "download url should start with base url"
        );
        debug_assert!(
            !fields.download_url.as_str().ends_with('/'),
            "download url should not end with trailing slash"
        );
        debug_assert!(
            fields.base_url.as_str().ends_with('/'),
            "base url should end with trailing slash"
        );

        self.handlebars
            .render("download_page", &fields)
            .expect("Failed to render download_page template")
    }
}

#[cfg(test)]
mod tests {
    use chrono::TimeZone;
    use insta::assert_snapshot;

    use super::*;

    #[test]
    fn test_render_download_page() {
        let templates = Templates::new();
        let fields = DownloadPageFields {
            file_name: "homework.docx".to_string(),
            username_source_url: Url::parse("https://www.github.com/josiahbull")
                .expect("hardcoded url to always be valid"),
            username: "josiahbull".to_string(),
            upload_date: chrono::Utc.timestamp_opt(1_620_000_000, 0).unwrap(),
            expiry_date: chrono::Utc.timestamp_opt(1_630_000_000, 0).unwrap(),
            file_size_bytes: NonZeroU64::new(1024).expect("file size should be non-zero"),
            download_url: Url::parse("https://www.downloads.mysite.com/homework.docx")
                .expect("hardcoded url to always be valid"),
            base_url: Url::parse("https://www.downloads.mysite.com/")
                .expect("hardcoded url to always be valid"),
        };
        let rendered = templates.render_download_page(fields);

        assert!(rendered.contains("homework.docx"));
        assert!(rendered.contains("josiahbull"));
        assert!(rendered.contains("1620000000"));
        assert!(rendered.contains("1024"));
        assert!(rendered.contains("https://www.downloads.mysite.com/homework.docx"));

        assert_snapshot!(rendered);
    }
}
