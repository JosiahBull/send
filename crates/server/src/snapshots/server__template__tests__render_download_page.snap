---
source: crates/server/src/template.rs
expression: rendered
snapshot_kind: text
---
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="description" content="Download page for a file named homework.docx.">
  <meta name="author" content="josiahbull">
  <meta name="keywords" content="download, file, send, share, cli, curl, wget">

  <meta name="darkreader-lock">

  <meta name="robots" content="noindex, nofollow" />

  <meta property="og:title" content="homework.docx Download">
  <meta property="og:description" content="Download page for homework.docx.">
  <meta property="og:image" content="https://www.downloads.mysite.com//favicon-192x192.png">
  <meta property="og:url" content="https://www.downloads.mysite.com/homework.docx">
  <meta property="og:type" content="website">

  <title>homework.docx</title>

  <link rel="icon shortcut" href="/favicon.ico" type="image/x-icon">
  <link rel="icon" href="/favicon.ico" sizes="16x16" type="image/x-icon">
  <link rel="icon" href="/favicon-32x32.png" sizes="32x32" type="image/png">
  <link rel="icon" href="/favicon-96x96.png" sizes="96x96" type="image/png">
  <link rel="icon" href="/favicon-192x192.png" sizes="192x192" type="image/png">
  <link rel="apple-touch-icon" href="/apple-touch-icon.png" sizes="180x180">

  <link rel="stylesheet" href="/styles.css">
</head>
<body data-theme="light">
  <button class="theme-toggle" onclick="toggleTheme()">Light Mode</button>

  <div class="main-content">
    <div class="container">
      <h1>Download File</h1>
      <div class="file-info">
        <p><strong>File Name:</strong> homework.docx</p>
        <p><strong>Uploaded By:</strong> <a href="https://www.github.com/josiahbull">josiahbull</a></p>
        <p><strong>Uploaded:</strong> <span class="uploadedAt">1620000000</span></p>
        <p><strong>Expires: </strong> <span class="expiresAt">1630000000</span></p>
        <p><strong>File Size:</strong> 1024</p>
      </div>
      <a href="https://www.downloads.mysite.com/homework.docx" class="button" download="homework.docx">Download via Browser</a>
      <div class="cli-instructions">
        <p><strong>Download via CLI:</strong></p>
        <div class="cli-command">
            <span class="code">curl -o homework.docx https://www.downloads.mysite.com/homework.docx</span>
          <svg class="copy-icon" onclick="copyCommand('curl -o homework.docx https://www.downloads.mysite.com/homework.docx')" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M19 3h-8c-1.1 0-2 .9-2 2v2H5c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2v-2h4c-1.1 0-2-.9 2-2V5c0-1.1-.9-2-2-2zm-6 14H5V9h8v8zm6-4h-4V5h4v8z"/>
          </svg>
        </div>
        <div class="cli-command">
            <span class="code">wget -O homework.docx https://www.downloads.mysite.com/homework.docx</span>
          <svg class="copy-icon" onclick="copyCommand('wget -O homework.docx https://www.downloads.mysite.com/homework.docx')" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M19 3h-8c-1.1 0-2 .9-2 2v2H5c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2v-2h4c-1.1 0-2-.9 2-2V5c0-1.1-.9-2-2-2zm-6 14H5V9h8v8zm6-4h-4V5h4v8z"/>
          </svg>
        </div>
      </div>
      <div class="upload-cta">
        <p>Want to upload a file? <a href="/">Click here to upload</a>.</p>
      </div>
    </div>
  </div>
  <div class="social-links">
    <a href="https://www.linkedin.com/in/josiah-bull">
      <img src="https://cdn.jsdelivr.net/npm/simple-icons/icons/linkedin.svg" alt="LinkedIn">
      LinkedIn
    </a>
    <a href="https://github.com/josiahbull/send">
      <img src="https://cdn.jsdelivr.net/npm/simple-icons/icons/github.svg" alt="GitHub">
      GitHub
    </a>
  </div>

  <div id="toast" class="toast"></div>

  <script>
    const uploadedAt = new Date(uploadedAtUtc);
    const expiresAt = new Date(expiresAtUtc);

    const uploadedAtElement = document.querySelector('.uploadedAt');
    const expiresAtElement = document.querySelector('.expiresAt');

    //
  </script>

  <!-- JavaScript for handling theme changes/toggles -->
  <script>
    const themeToggle = document.querySelector('.theme-toggle');
    const root = document.documentElement;

    const themes = ['light', 'dark', 'auto'];
    let currentTheme = localStorage.getItem('theme') || 'auto';

    function setTheme(theme) {
      if (theme === 'auto') {
        const prefersDark = window.matchMedia('(prefers-color-scheme: dark)').matches;
        theme = prefersDark ? 'dark' : 'light';
      }
      document.body.setAttribute('data-theme', theme);
      themeToggle.textContent = `${capitalize(currentTheme)} Mode`;
      localStorage.setItem('theme', currentTheme);
    }

    function toggleTheme() {
      const currentIndex = themes.indexOf(currentTheme);
      currentTheme = themes[(currentIndex + 1) % themes.length];
      setTheme(currentTheme);
    }

    function capitalize(word) {
      return word.charAt(0).toUpperCase() + word.slice(1);
    }

    // Apply theme on page load
    setTheme(currentTheme);

    // Update theme if system setting changes and in Auto Mode
    window.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', () => {
      if (currentTheme === 'auto') setTheme('auto');
    });
  </script>

  <!-- JavaScript for handling toast notification and copying command -->
  <script>
    function copyCommand(command) {
      navigator.clipboard.writeText(command).then(() => {
        showToast(`Command copied: ${command}`);
      }).catch(err => {
        showToast(`Failed to copy: ${err}`);
      });
    }

    function showToast(message) {
      const toast = document.getElementById('toast');
      toast.textContent = message;
      toast.classList.add('show');
      setTimeout(() => {
        toast.classList.remove('show');
      }, 3000);
    }
  </script>
</body>
</html>
